lookup_optim.go

```diff
@@ -0,0 +1,306 @@
+	putProvDone atomic.Int32
+}
+
+func (dht *IpfsDHT) newOptimisticState(ctx context.Context, key string) (*optimisticState, error) {
+	// get network size and err out if there is no reasonable estimate
+	networkSize, err := dht.nsEstimator.NetworkSize()
+	if err != nil {
+		return nil, err
+	}
+
+	individualThreshold := mathext.GammaIncRegInv(float64(dht.bucketSize), 1-OptProvIndividualThresholdCertainty) / float64(networkSize)
+	setThreshold := mathext.GammaIncRegInv(float64(dht.bucketSize)/2.0+1, 1-OptProvSetThresholdStrictness) / float64(networkSize)
+	returnThreshold := int(math.Ceil(float64(dht.bucketSize) * OptProvReturnRatio))
+
+	return &optimisticState{
+		putCtx:              ctx,
+		dht:                 dht,
+		key:                 key,
+		doneChan:            make(chan struct{}, returnThreshold), // buffered channel to not miss events
+		ksKey:               ks.XORKeySpace.Key([]byte(key)),
+		networkSize:         networkSize,
+		peerStates:          map[peer.ID]addProviderRPCState{},
+		individualThreshold: individualThreshold,
+		setThreshold:        setThreshold,
+		returnThreshold:     returnThreshold,
+		putProvDone:         atomic.Int32{},
+	}, nil
+}
+
+func (dht *IpfsDHT) optimisticProvide(outerCtx context.Context, keyMH multihash.Multihash) error {
+	key := string(keyMH)
+
+	if key == "" {
+		return fmt.Errorf("can't lookup empty key")
+	}
+
+	// initialize new context for all putProvider operations.
+	// We don't want to give the outer context to the put operations as we return early before all
+	// put operations have finished to avoid the long tail of the latency distribution. If we
+	// provided the outer context the put operations may be cancelled depending on what happens
+	// with the context on the user side.
+	putCtx, putCtxCancel := context.WithTimeout(context.Background(), time.Minute)
+
+	es, err := dht.newOptimisticState(putCtx, key)
+	if err != nil {
+		putCtxCancel()
+		return err
+	}
+
+	// initialize context that finishes when this function returns
+	innerCtx, innerCtxCancel := context.WithCancel(outerCtx)
+	defer innerCtxCancel()
+
+	go func() {
+		select {
+		case <-outerCtx.Done():
+			// If the outer context gets cancelled while we're still in this function. We stop all
+			// pending put operations.
+			putCtxCancel()
+		case <-innerCtx.Done():
+			// We have returned from this function. Ignore cancellations of the outer context and continue
+			// with the remaining put operations.
+		}
+	}()
+
+	lookupRes, err := dht.runLookupWithFollowup(outerCtx, key, dht.pmGetClosestPeers(key), es.stopFn)
+	if err != nil {
+		return err
+	}
+
+	// Store the provider records with all the closest peers
+	// we haven't already contacted.
+	es.peerStatesLk.Lock()
+	for _, p := range lookupRes.peers {
+		if _, found := es.peerStates[p]; found {
+			continue
+		}
+
+		go es.putProviderRecord(p)
+		es.peerStates[p] = Sent
+	}
+	es.peerStatesLk.Unlock()
```

Comment:
Sorry I'm finding the query code very difficult to understand. What are the scenarios where a peer in `lookupRes.peers` was not already in the qps passed to the `stopFn` during the query?

Reply:
`lookupRes.peers` will always only contain the peers that were already present in the qps. `lookupRes.peers` will contain the 20 closest peers that we found at the end of `runLookupWithFollowup` (subset of all peers we have in qps).

In the context of optimistic provide: During `runLookupWithFollowup` we have attempted to store provider records with peers while we're still looking for closer peers. After `runLookupWithFollowup` returns (aka we think we have found the closest), we loop through the returned peers and only contact the ones that we haven't previously contacted (during `runLookupWithFollowup` while we were still looking for closer peers).